cpp/src/arrow/flight/server.cc

```diff
@@ -293,7 +209,36 @@ class FlightServiceImpl : public FlightService::Service {
 
   grpc::Status DoPut(ServerContext* context, grpc::ServerReader<pb::FlightData>* reader,
                      pb::PutResult* response) {
-    return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "");
+    // Get metadata
+    pb::FlightData data;
+    if (reader->Read(&data)) {
+      FlightDescriptor descriptor;
+      std::unique_ptr<ipc::Message> message;
+      std::shared_ptr<Schema> schema;
+      GRPC_RETURN_NOT_OK(internal::FromProto(data.flight_descriptor(), &descriptor));
+      std::shared_ptr<Buffer> header_buf;
+      std::shared_ptr<Buffer> body_buf;
+      GRPC_RETURN_NOT_OK(Buffer::FromString(data.data_header(), &header_buf));
+      GRPC_RETURN_NOT_OK(Buffer::FromString(data.data_body(), &body_buf));
+      GRPC_RETURN_NOT_OK(ipc::Message::Open(header_buf, body_buf, &message));
+
+      if (!message || message->type() != ipc::Message::Type::SCHEMA) {
+        return internal::ToGrpcStatus(
+            Status(StatusCode::Invalid, "DoPut must start with schema/descriptor"));
+      } else {
+        GRPC_RETURN_NOT_OK(ipc::ReadSchema(*message, &schema));
+
+        auto message_reader = std::unique_ptr<FlightMessageReader>(
+            new FlightMessageReaderImpl(descriptor, schema, reader));
+        return internal::ToGrpcStatus(server_->DoPut(std::move(message_reader)));
+      }
+    } else {
+      // TODO(lihalite): gRPC doesn't let us distinguish between no
+      // message sent, and message failed to deserialize. IMO, we
+      // should add logging around the Status returns in
+      // serialization-internal.h to make debugging such cases easier.
+      return grpc::Status::OK;
```

Comment:
Yes, though in this particular case the `SerializationTraits` customization isn't being invoked unless I misread. Does gRPC provide any way to distinguish deserialization failures in its out-of-the-box protobuf API?

Reply:
Ah, I think I changed it at some point, but I did run into the problem described before. I'll take a closer look at their APIs.

Reply:
Hmm, can you explain why you return OK here? If the client doesn't send a first message, then probably it should be considered an IPC failure?
(for example, the client might have called `WritesDone` but still be available for receiving a response... am I misunderstanding this?)

Reply:
Hmm, that's fair. I'll change it to gRPC INVALID_ARGUMENT.